packages/edit-site/src/components/navigation-sidebar/navigation-panel/use-debounced-search.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useCallback, useMemo, useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { SEARCH_DEBOUNCE_IN_MS } from './constants';
+
+export default function useDebouncedSearch() {
+	const [ searchInput, setSearchInput ] = useState( '' );
+	const [ search, setSearch ] = useState( '' );
+	const [ debouncing, setDebouncing ] = useState( false );
+
+	useEffect( () => {
+		setDebouncing( false );
+	}, [ search ] );
+
+	const debouncedSetSearch = useCallback(
+		debounce( setSearch, SEARCH_DEBOUNCE_IN_MS ),
+		[ setSearch ]
+	);
+
+	const handleOnSearch = useCallback(
+		( value ) => {
+			setSearchInput( value );
+			debouncedSetSearch( value );
+			setDebouncing( true );
+		},
+		[ setSearchInput, debouncedSetSearch ]
+	);
+
+	const menuProps = useMemo(
+		() => ( {
+			hasSearch: true,
+			onSearch: handleOnSearch,
+			search: searchInput,
+		} ),
+		[ handleOnSearch, searchInput ]
+	);
+
+	return { search, debouncing, menuProps };
```

Comment:
I feel there's one naming switcheroo too many that makes it slightly unclear what variable does what. 🤔 

What about something like this?

```jsx
function useDebouncedSearch() {
	// The value used by the NavigationMenu to control the input field.
	const [ search, setSearch ] = useState( '' );
	// The value used to actually perform the search query.
	const [ searchQuery, setSearchQuery ] = useState( '' );
	const [ isDebouncing, setIsDebouncing ] = useState( false );

	useEffect( () => {
		setIsDebouncing( false );
	}, [ searchQuery ] );

	const debouncedSetSearchQuery = debounce(
		setSearchQuery,
		SEARCH_DEBOUNCE_IN_MS
	);

	const onSearch = ( value ) => {
		setSearch( value );
		debouncedSetSearchQuery( value );
		setIsDebouncing( true );
	};

	return { isDebouncing, onSearch, search, searchQuery };
}
```

In the example I've ignored the `useCallback` and `useMemo` for clarity, but also because I wonder if they actually give us appreciable performance gains. 🤔